src/odemis/driver/keysight.py

```diff
@@ -122,144 +187,259 @@ def _sendCommand(self, cmd):
+        self._checkError()
 
     def _set_period(self, p: float) -> float:
-        f = 1/p
+        f = 1 / p
         self.setFrequency(self._channel, f)
-        self.delay.value = min(self.delay.value, p)
-        return p
-
-    def _set_delay(self, d):
-        if not (0 <= d <= self.period.value):  #% (bs,)
+        # Update the delay (aka phase), to match in terms of time, and stay within the period
+        try:
+            delay_max = p * 0.99999  # tiny bit less than max, as the device complains if the delay is exactly the period
+            delay_clipped = min(max(-delay_max, self.delay.value), delay_max)
+            if delay_clipped != self.delay.value:
+                self.delay.value = delay_clipped
+        except TrueFormError:
+            logging.warning("Could not set the delay to match the period")
+
+        actual_f = self.getFrequency(self._channel)
+        return 1 / actual_f
+
+    def _set_delay(self, d: float) -> float:
+        if not (-self.period.value <= d <= self.period.value):
             raise ValueError("Delay must be < period (%s), got %s" % (self.period.value, d))
 
         self.setPhase(self._channel, d)
-        return d
+        act_phase = self.getPhase(self._channel)
+        act_d = self.period.value * act_phase / 360
+        return act_d
 
-    def _set_duty_cycle(self, dc):
+    def _set_duty_cycle(self, dc: float) -> float:
         duty_cycle = dc * 100
-        logging.debug("%r ,%s", duty_cycle, type(duty_cycle))
-        self.setDutyCycle(self._channel, duty_cycle)
-        return dc
+        exp = None
+        try:
+            self.setDutyCycle(self._channel, duty_cycle)
+        except (IndexError, TrueFormError) as e:
+            exp = e
+
+        act_dc = self.getDutyCycle(self._channel) / 100
+        # In case of error, the value still might have changed, so need to update it
+        if exp:
+            self.dutyCycle._set_value(act_dc)
+            raise exp
+
+        return act_dc
 
     def _set_power(self, p: bool) -> bool:
-        c = self._channel
-        if p:
-            lim = self._limits[c - 1]
-            self.setVoltageMin(c, lim[0])
-            self.setVoltageMax(c, lim[1])
-            self.setOutput(c, True)
-        else:
-            # set min and max voltage limits to prevent possible glitch issues
-            self.setVoltageMin(c, 0)
-            self.setVoltageMax(c, 0.01)
-            self.setOutput(c, False)
+        # Note: when power is off, the sync is still sent, but not waveform signal is put back to 0V.
+        self.setOutput(self._channel, p)
 
         for c in self._tracking.keys():
             self.setOutput(c, p)
 
-        return p
+        act_p = self.getOutput(self._channel)
+        return act_p
+
+    def _update_settings(self):
+        """
+        Update the settings of the device.
+        Called by the repeating timer.
+        """
+        f = self.getFrequency(self._channel)
+        p = 1 / f
```

Comment:
Will the frequency ever be `0`?

Reply:
No, the device doesn't accept a frequency of 0, and never returns such value. I'll add a comment.